Scala 3

ybasket opened the pull request:
Builds on top of #228, so the changes from there are included here as well.

Adds cross build for Scala 3 which is a bit tricky due to:
* usage of shapeless' `Sized`. Luckily, the API surface used is very small, so I was able to "polyfill" it by defining some match types which mimic `Sized`, but represent the data as tuples internally (which would be the natural choice if this library was developed exclusively for Scala 3).
* Custom macros. They were easy to port, but as Scala 2 doesn't allow macro calls from a `trait`, the companion objects using macros for their `apply` methods had be moved into Scala version-specific files. That works well, but tears the code a bit apart. Striving for as little changes as possible (tests actually compile on Scala 2 unchanged), this was the only solution I found. If in future the `Name.apply` and friends would vanish in favour of adopting Literally, this could become nicer again.

I'll comment on some interesting parts inline later.

ybasket commented:
The tests pass on my fork, so this version of minimal API surface changes is ready for review.

If the decision for Literally (which wouldn't work exactly straight-forward for `Quantile`) was a thing, I would update this PR instead of following up with another PR as much code would be moved back and forth. @ChristopherDavenport you mentioned Literally for this project on Discord some time ago, is that still something you want?

ChristopherDavenport commented:
1. The case class additions are problematic from a type exposure perspective, would love if we could do without those.
2. Literally works for me but I'm happy to proceed with this.

ybasket commented:
> 1. The case class additions are problematic from a type exposure perspective, would love if we could do without those.
> 2. Literally works for me but I'm happy to proceed with this.

I fixed 1. by adding a `private[epimetheus] def asJava` method on each collector type. Also not ideal IMHO, but trying to pattern match in different ways always leads to compiler errors, especially on Scala 2.12 – that was the reason I introduced `case` initially. @ChristopherDavenport I hope this work around is acceptable, other options would be to drop 2.12 or try manually spell out a private `unapply` for each class and hope compilers are happy about it.

2. As Literally isn't a perfect fit (`Quantile`), I would then not migrate, at least not as part of this PR.
